src/ecephys_etl/modules/vbn_create_stimulus_table/create_stim_table.py

```diff
@@ -0,0 +1,551 @@
+    tolerance : int
+        Percent by which frame counts are allowed to deviate from,
+        by default 0.
+
+    Returns
+    -------
+    List[float]
+        List of the inferred start frames for each stimuli category presented
+        during session.
+    """
+    frame_count_arr = np.array(frame_counts)
+    tolerance_pct = tolerance / 100.
+
+    # Get vsyncs and stim_running signals from sync data
+    vsync_times = get_vsyncs(sync_dataset)
+    stim_starts, stim_ends = get_stim_starts_and_ends(sync_dataset)
+
+    # Get vsync frame lengths and indices for all stimuli
+    epoch_frame_counts = []
+    epoch_start_frames = []
+    for start, end in zip(stim_starts, stim_ends):
+        # Inner expression returns a bool array where conditions are True
+        # np.where evaluates bool array to return indices where bool array True
+        epoch_frames = np.where((vsync_times > start) & (vsync_times < end))[0]
+        epoch_frame_counts.append(len(epoch_frames))
+        epoch_start_frames.append(epoch_frames[0])
+
+    if len(epoch_frame_counts) == len(frame_count_arr):
+        if not np.allclose(frame_count_arr, epoch_frame_counts):
+            logging.warning(
+                f"Number of frames derived from sync file "
+                f"({epoch_frame_counts})for each epoch not matching up with "
+                f"frame counts derived from pkl files ({frame_count_arr})!"
+            )
+        start_frames = epoch_start_frames
+    elif len(epoch_frame_counts) > len(frame_count_arr):
+        logging.warning(
+            f"Number of stim presentations obtained from sync "
+            f"({len(epoch_frame_counts)}) higher than number expected "
+            f"({len(frame_count_arr)}). Inferring start frames."
+        )
+
+        start_frames = []
+        for stim_idx, fc in enumerate(frame_count_arr):
+
+            logging.info(f"Finding stim start for stim with index: {stim_idx}")
+            # Get index of stimulus whose frame counts most closely match
+            # the expected number of frames
+            best_match = int(
+                np.argmin([np.abs(efc - fc) for efc in epoch_frame_counts])
+            )
+            lower_tol = fc * (1 - tolerance_pct)
+            upper_tol = fc * (1 + tolerance_pct)
+            if lower_tol <= epoch_frame_counts[best_match] <= upper_tol:
+                _ = epoch_frame_counts.pop(best_match)
+                start_frame = epoch_start_frames.pop(best_match)
+                start_frames.append(start_frame)
+                logging.info(
+                    f"Found stim start for stim with index ({stim_idx})"
+                    f"at vsync ({start_frame})"
+                )
+            else:
+                raise RuntimeError(
+                    f"Could not find matching sync frames for stim: {stim_idx}"
+                )
+    else:
+        raise RuntimeError(
+            f"Do not know how to handle more pkl frame count entries "
```

Comment:
@corbennett The original implementation only addressed when `len(epoch_frame_counts) > len(frame_count_arr`. Does there need to be code to address the `<` case?

Reply:
The < case implies a sync data stream failure (or a mismatch between the experiment generating the pkl files and the experiment generating the sync file). Raising an error for this as you have done here makes sense.